third_party/amd/lib/TritonAMDGPUTransforms/UpdateAsyncWaitCount.cpp

```diff
@@ -0,0 +1,139 @@
+#include "amd/lib/TritonAMDGPUToLLVM/Utility.h"
+#include "amd/lib/TritonAMDGPUTransforms/Utility.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+#define GEN_PASS_CLASSES
+#include "TritonAMDGPUTransforms/Passes.h"
+
+using namespace mlir;
+namespace tt = triton;
+namespace ttg = triton::gpu;
+
+// LLVM cannot infer the dependency between direct to lds (async) loads and the
+// local reads between warps in a workgroup. As a workaround we update the
+// waitcnt to represent the number of hardware instructions we are interleaving
+// with. This allows us to manually emit the waitcnt during lowering.
+struct UpdateAsyncWaitCount : public OpRewritePattern<ttg::AsyncWaitOp> {
+  UpdateAsyncWaitCount(MLIRContext *ctx) : OpRewritePattern(ctx) {}
+
+  LogicalResult matchAndRewrite(ttg::AsyncWaitOp waitOp,
+                                PatternRewriter &rewriter) const override {
+    int waitCnt = std::numeric_limits<int>::max();
+
+    // AsyncWait can await multiple tokens so we get the minimum from all
+    // tokens
+    for (auto token : waitOp.getOperands()) {
+      // Traverse def chain from waitOp to the producer of the token and count
+      // the minumum number of vmcnt instructions
+      auto tokenWaitCnt =
+          findMinPathCountInDefChain(token, waitOp, [this](Operation *op) {
+            return getNumberOfLoadInstructions(op);
+          });
+      waitCnt = std::min(waitCnt, tokenWaitCnt);
+    }
+
+    if (waitCnt == std::numeric_limits<int>::max() ||
+        waitOp.getNum() == waitCnt)
+      return failure();
+
+    rewriter.modifyOpInPlace(waitOp, [&]() { waitOp.setNum(waitCnt); });
+    return success();
+  }
+
+  // To count AsyncLoads we have two options:
+  // 1) Count the AsyncLoads directly but we the overall producer of the token
+  //    might be an AsyncCommitGroup in which case we would miss the loads of
+  //    this CommitGroup.
+  // 2) Count based on the Operands of encountered CommitGroups. This ensures we
```

Comment:
I think this comment can be revised to be clear if we follow a structure like "Pipeliner always insert ops following an order of ttg.async_load -> [token] -> ttg.async_commit_group -> [token] -> ttg.async_wait. So here we scan the operands of ttg.async_commit_group to count the number of issued async load intrinsics. ..."